Source/engine/render/text_render.cpp

```diff
@@ -309,16 +234,32 @@ uint16_t DrawString(const Surface &out, string_view text, const Rectangle &rect,
 	if (lineHeight == -1)
 		lineHeight = LineHeights[size];
 
-	uint16_t i = 0;
+	if (HasAnyOf(flags, UiFlags::VerticalCenter)) {
+		int textHeight = (std::count(text.cbegin(), text.cend(), '\n') + 1) * lineHeight;
+		characterPosition.y += (rect.size.height - textHeight) / 2;
+	}
+
+	characterPosition.y += BaseLineOffset[size];
+
+	uint32_t fontId = (color << 24) | (size << 16);
+	auto font = Fonts.find(fontId);
+	if (font == Fonts.end()) {
+		Log("Font: size {} and color {} not loaded ", size, color);
+		return 0;
+	}
+
+	const auto &activeFont = font->second;
+
+	uint32_t i = 0;
 	for (; i < text.length(); i++) {
-		uint8_t frame = FontFrame[size][FontIndex[static_cast<uint8_t>(text[i])]];
-		int symbolWidth = FontKern[size][frame];
+		uint8_t frame = text[i] & 0xFF;
+		int symbolWidth = FontKerns[size][frame];
 		if (text[i] == '\n' || characterPosition.x + symbolWidth > rightMargin) {
 			if (characterPosition.y + lineHeight >= bottomMargin)
 				break;
 			characterPosition.y += lineHeight;
 
-			if (HasAnyOf(flags, (UiFlags::AlignCenter | UiFlags::AlignRight | UiFlags::KerningFitSpacing)))
+			if (HasAnyOf(flags, (UiFlags::AlignCenter | UiFlags::AlignRight | UiFlags::KerningFitSpacing)) && i < text.length())

```

Comment:
```suggestion
			if (HasAnyOf(flags, (UiFlags::AlignCenter | UiFlags::AlignRight | UiFlags::KerningFitSpacing)) && i + 1 < text.length())
```

Reply:
This fixes the OOB error, but is the logic correct? I understand skipping past the character to calculate `lineWidth` in the case of encountering a newline. However, when a character's width exceeds the right margin, that character gets moved down to the next line so shouldn't we then include that character in the `lineWidth` calculation?

Reply:
Maybe something like this.
```suggestion
			if (HasAnyOf(flags, (UiFlags::AlignCenter | UiFlags::AlignRight | UiFlags::KerningFitSpacing))) {
				int nextLineIndex = text[i] == '\n' ? i + 1 : i;
				if (nextLineIndex < text.length())
					lineWidth = GetLineWidth(&text[nextLineIndex], size, spacing, &charactersInLine);
				else
					lineWidth = 0;
			}
```

Reply:
Good catch, I think I'll actually just simplify the logic here instead. The original problem with rendering past the edge of the box is largely alleviated by the back-buffer now being clipped so that content doesn't wrap the screen.

Reply:
With the latest change, the OOB can only occur if a center-aligned, right-aligned, or kerning-adjusted string ends in a newline. In the first place, there aren't any strings in the game that end in newlines so for now it should be okay.